orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -106,6 +102,26 @@ public ArtifactResolver(ObjectMapper objectMapper, ExecutionRepository execution
     return emittedArtifacts;
   }
 
+  /**
+   * Used to fully resolve a bound artifact on a stage that can either select
+   * an expected artifact ID for an expected artifact defined in a prior stage
+   * or as a trigger constraint OR define an inline expression-evaluable default artifact.
+   * @param stage The stage containing context to evaluate expressions on the bound artifact.
+   * @param id An expected artifact id. Either id or artifact must be specified.
+   * @param artifact An inline default artifact. Either id or artifact must be specified.
+   * @return A bound artifact with expressions evaluated.
+   */
+  public @Nullable Artifact getBoundArtifactForStage(Stage stage, @Nullable String id, @Nullable Artifact artifact) {
+    Artifact boundArtifact = id != null ? getBoundArtifactForId(stage, id) : artifact;
```

Comment:
Oh, it necessary to evaluate the passed in Artifact too?

Reply:
Yes.

Reply:
Because the passed in Artifact will come from an inline default artifact which may have expressions:

![image](https://user-images.githubusercontent.com/1697736/53893587-19447e00-3ff4-11e9-99c6-bb22e4491941.png)